### ISSUE TITLE
Comparing blobs

### ISSUE THREAD

**Issue by prospectumdev** [fnc12/sqlite_orm#1028]
Hi,

I am using a blob `vector<char> hoi_ids`;

Now when I perform this query...

`auto existingTestCases = x->get_all<db::TestCase>(where(c(&db::TestCase::pdOrganParameters_id) == tc.pdOrganParameters_id and c(&db::TestCase::pdRankParameters_id)  == tc.pdRankParameters_id and c(&db::TestCase::hoi_ids) == tc.hoi_ids));`

...the query created by sqlite_orm looks like this...

```
SELECT 'TestCase'."id", 'TestCase'."pdOrganParameters_id", 'TestCase'."pdRankParameters_id", 'TestCa
se'."hoi_ids", 'TestCase'."exec_time", 'TestCase'."successRate", 'TestCase'."failedMIDIkeys" FROM 'T
estCase' WHERE (  ( ('TestCase'."pdOrganParameters_id" = 1) AND ('TestCase'."pdRankParameters_id" =
1001) ) AND ('TestCase'."hoi_ids" = ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺
☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺
☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺
☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺
☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺
☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ☺   ) )
```

...and in line 8405 of sqlite_orm.h an exception with error code SQLITE_MISUSE_BKPT (sqlite3.c, line 157680) is thrown.

The problem is the comparison `c(&db::TestCase::hoi_ids) == tc.hoi_ids`.

So what is the proper way to compare blobs within a sqlite_orm query?

Thanks!

**Comment by fnc12** [comment]
what is your `hoi_ids` vector type?

**Comment by trueqbit** [comment]
What version of sqlite_orm are you using? @prospectumdev As a programmer you must make a proper problem report, please.
I presume that your query works with the current dev branch.

**Comment by prospectumdev** [comment]
I am using version 1.6.0

**Comment by prospectumdev** [comment]
> what is your `hoi_ids` vector type?

`vector<char>`

**Comment by trueqbit** [comment]
Are you able to reproduce this in a test program? I am able to do a select with a where clause on a vector<char>. The documentation for `SQLITE_MISUSE` states that a program that encounters this error might be ill-formed in other places.

In regard to the sql string: not sure how you produced it, but can you look at what a prepared statement returns? i.e.:
```c++
auto statement = x->prepare(get_all<db::TestCase>(where(c(&db::TestCase::hoi_ids) == tc.hoi_ids)));
std::string sql = statement.sql();
auto results = x->execute(statement);
```

**Comment by prospectumdev** [comment]
Here is a little test program:

```
namespace db
{
	struct CompTest
	{
		int id = -1;
		vector<char> blob;
	};

	auto initMyCompTest(const string& path)
	{
		return make_storage(path,
			make_table
			(
				"compTest",
				make_column("id", &CompTest::id, autoincrement(), primary_key(), unique()),
				make_column("blob", &CompTest::blob)
			)
		);
	}

	using MyCompTest = decltype(initMyCompTest(""));
}

void blobComparisonTest()
{
	{
		db::MyCompTest myCompTest(db::initMyCompTest("comptest.sqlite3"));
		myCompTest.sync_schema();

		//create test data
		db::CompTest temp({ 1, vector<char>(128, 0) });
		myCompTest.replace(temp);

		//comparison test
		try 
		{
			auto existingTestCases = myCompTest.get_all<db::CompTest>(where(c(&db::CompTest::blob) == temp.blob));
		}
		catch (const std::system_error& e)
		{
			std::cout << "Error '" << e.code() << "' meaning '" << e.what() << "'\n";
		}

		//prepared statement test
		auto statement = myCompTest.prepare(get_all<db::CompTest>(where(c(&db::CompTest::blob) == temp.blob)));
		std::string sql = statement.sql();
		auto results = myCompTest.execute(statement);
		std::cout << "Prepared statement test result:\n" << sql << "'\n";
	}
}
```

An interesting observation is that **prepare() and execute() seem to be unavailable** so that the 'prepared statement test' must be commented out in order for the program to compile. Then, I get this error message

> Error 'SQLite error:1' meaning 'SQL logic error'

**Comment by prospectumdev** [comment]
Ok, I am now using v1.7.1.. Had to add `#define SQLITE_ORM_OMITS_CODECVT` to get it compiled.

Now, `auto existingTestCases = myCompTest.get_all<db::CompTest>(where(c(&db::CompTest::blob) == temp.blob));` is
an array of size 1, as it should be.

Prepared statements also work again:

> Prepared statement test result:
> SELECT "compTest"."id", "compTest"."blob" FROM 'compTest'  WHERE ( ("compTest"."blob" = ?)) '

Now, `auto results = myCompTest.execute(statement);` is an array of size 1, as it should be.

That looks promising! Seems that I have stumbled over an issue that was there in an older version but has already been fixed.
What I'll do now is go back to my initial problem, check if it has disappeared, too, and get back to you for a (most likely) final reply....

and yes, my code works now as expected! Thank you very much!
